packages/lesswrong/server/resolvers/analyticsResolvers.ts

```diff
@@ -161,6 +177,186 @@ addGraphQLResolvers({
+          slug: 1,
+          postedAt: 1,
+          baseScore: 1,
+          commentCount: 1,
+        },
+      }).fetch();
+
+      const postsById = Object.fromEntries(userPosts.map((post) => [post._id, post]));
+      const postIds = userPosts.map((post) => post._id);
+
+      if (!postIds.length) {
+        return {
+          posts: [],
+          totalCount: 0,
+        };
+      }
+
+      const postViewsView = getHybridView(POST_VIEWS_IDENTIFIER);
+      const postViewTimesView = getHybridView(POST_VIEW_TIMES_IDENTIFIER);
+
+      if (!postViewsView || !postViewTimesView) throw new Error("Hybrid views not configured");
+
+      const [viewsTable, postViewTimesTable] = await Promise.all([
+        postViewsView.virtualTable(),
+        postViewTimesView.virtualTable(),
+      ]);
+
+      const postIdsBatches = chunk(postIds, BATCH_SIZE);
+
+      let viewsResults: { _id: string; total_view_count: number }[] = [];
+      let readsResults: { _id: string; total_read_count: number }[] = [];
+
+      async function runViewsQuery(batch: string[]) {
+        const viewsResult = await analyticsDb.any<{ _id: string; total_view_count: number }>(
+          `
+          SELECT
+            post_id AS _id,
+            sum(view_count) AS total_view_count
+          FROM
+            (${viewsTable}) q
+          WHERE
+            ${generateOrConditionQuery("post_id", batch)}
+          GROUP BY
+            post_id;
+        `,
+          batch
+        );
+
+        viewsResults.push(...viewsResult);
+      }
+
+      async function runReadsQuery(batch: string[]) {
+        const readsResult = await analyticsDb.any<{ _id: string; total_read_count: number }>(
+          `
+          SELECT
+            post_id AS _id,
+            -- A "read" is anything with a reading_time over 30 seconds
+            sum(CASE WHEN reading_time > 30 THEN 1 ELSE 0 END) AS total_read_count
+          FROM
+            (${postViewTimesTable}) q
+          WHERE
+            ${generateOrConditionQuery("post_id", batch)}
+          GROUP BY
+            post_id;
+        `,
+          batch
+        );
+
+        readsResults.push(...readsResult);
+      }
+
+      // Pair batches of ids with their corresponding query functions and then zip these pairs together
+      const zippedPairs = postIdsBatches.flatMap((batch) => [
+        { batch, func: runViewsQuery },
+        { batch, func: runReadsQuery },
+      ]);
+
+      let queue: Promise<any>[] = [];
+
+      // Execute up to maxConcurrent queries at a time
+      for (const { batch, func } of zippedPairs) {
+        // Add the new promise to the queue
+        const promise = func(batch);
+        queue.push(promise);
+
+        // If the queue is full, wait for one promise to finish
+        if (queue.length >= MAX_CONCURRENT_QUERIES) {
+          await Promise.race(queue);
+          // Remove resolved promises from the queue
+          queue = queue.filter((p) => inspect(p).includes("pending"));
```

Comment:
This seems a bit sketchy to me? GPT-4 suggests having each promise remove itself from the queue, which seems better to me:

```
// When a task finishes, remove it from the in-progress list and store its result
promise.then((result) => {
  inProgress = inProgress.filter(p => p !== promise);
  results.push(result);
});
```

Reply:
I got into an argument with GPT-4 about this... I think this is better because:
 - it's easier to follow what's going on because only the outer thread ever adds or removes promises from the queue, so you know the promises are just doing their task and the outer thread is controlling this queue behaviour
 - it ensures the promises are executed in the order they are given (not actually required in this case, but good for debugging, and more complicated to guaranteed if the promises control their own state)

The thing that is sketchy is that `util.inspect` is not completely guaranteed to keep supporting checking promise states, but I don't really see why it wouldn't

Reply:
Based on the documentation for `util.inspect()`, I really don't think we should be relying on it. Not that there's an imminent problem, but it just doesn't feel worth the risk. Also I don't think the current version ensures that promises are executed in the order they are given? I guess I don't think we gain anything by doing it the current way

```
* The `util.inspect()` method returns a string representation of `object` that is
* intended for debugging. The output of `util.inspect` may change at any time
* and should not be depended upon programmatically.
...
* `util.inspect()` is a synchronous method intended for debugging. Its maximum
* output length is approximately 128 MB. Inputs that result in longer output will
* be truncated. 
```